### ISSUE TITLE
Implement Reporter for Console JVM 

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#1229]
This ticket involves working from @NeQuissimus's [pull request](https://github.com/zio/zio/pull/723) (in particular, [here](https://github.com/zio/zio/pull/723/files)) to implement a `Reporter` for JVM console. The `Reporter` is essentially a function that takes an executed spec and reports it.

It would loop over an executed spec and print the classic list hierarchy, with color-coded text:

```
+ Addition works fine
+ Subtraction works fine
- Value falls within range
  `52` did not satisfy `(equals(42) || (gt(5) && lt(10))`
  `52` did not satisfy `gt(5)`
```

The first failure is the failure of the whole value / whole predicate; the second is the failure of the fragment value / fragment predicate.

The predicate fragment `gt(5)` inside the whole predicate would be colored differently, to draw user's attention to the fragment that failed. This can be accomplished by looking for substrings in the rendered predicate that match the rendered fragment predicate.

Failed test labels should be red, marked with a `-`; succeeded test labels should be marked with a green `+`.

<img width="537" alt="Screen Shot 2019-07-24 at 9 54 49 AM" src="https://user-images.githubusercontent.com/156745/61799441-1f979300-adf9-11e9-8aaf-afacdc8fb0a5.png">

**Notes**

 - See `testkit` `zio/test` for the source code for this ticket

**Comment by vasilmkd** [comment]
I'd like to take this if I may.

**Comment by adamgfraser** [comment]
I’m happy to take this.

**Comment by NeQuissimus** [comment]
I would not render the succeeded test labels green. Just leave them default and render the + in green, just like the screenshot.

**Comment by adamgfraser** [comment]
Sounds good.

**Comment by NeQuissimus** [comment]
Oh, and this needs to have an option to disable colours altogether. Otherwise the output is going to look like crap in some CI

**Comment by jdegoes** [comment]
@adamgfraser Fantastic!

**Comment by jdegoes** [comment]
It would be nice to do a summary of failures:

<img width="540" alt="Screen Shot 2019-07-27 at 2 12 38 PM" src="https://user-images.githubusercontent.com/156745/61998120-1b79a880-b079-11e9-8519-67b45eab0d96.png">

In some test frameworks, you have to scroll back to find failures; there is no summary at the end.

**Comment by NeQuissimus** [comment]
If you properly integrate with SBT and `test-framework`, it should give you a summary at the very end

**Comment by jdegoes** [comment]
@NeQuissimus Right, but it's terrible, it only tells you the class that failed and how many failures.

**Comment by NeQuissimus** [comment]
Be careful with a bounding box though, if people have long test names or tons of predicates, it's going to look awful ;)

**Comment by jdegoes** [comment]
True. Maybe better if it wraps. Or maybe just skip it for v1.

**Comment by ghostdogpr** [comment]
Closed by #1273
